esp-bootloader-support-esp-idf/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+    pub fn max_efuse_blk_rev_full(&self) -> u16 {
+        self.max_efuse_blk_rev_full
+    }
+
+    /// MMU page size
+    pub fn mmu_page_size(&self) -> u8 {
+        self.mmu_page_size
+    }
+}
+
+impl core::fmt::Debug for EspAppDesc {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("EspAppDesc")
+            .field("magic_word", &self.magic_word)
+            .field("secure_version", &self.secure_version)
+            .field("version", &self.version())
+            .field("project_name", &self.project_name())
+            .field("time", &self.time())
+            .field("date", &self.date())
+            .field("idf_ver", &self.idf_ver())
+            .field("app_elf_sha256", &self.app_elf_sha256)
+            .field("min_efuse_blk_rev_full", &self.min_efuse_blk_rev_full)
+            .field("max_efuse_blk_rev_full", &self.max_efuse_blk_rev_full)
+            .field("mmu_page_size", &self.mmu_page_size)
+            .finish()
+    }
+}
+
+fn array_to_str(array: &[core::ffi::c_char]) -> &str {
+    let len = array.iter().position(|b| *b == 0).unwrap_or(array.len());
+    unsafe {
+        core::str::from_utf8_unchecked(core::slice::from_raw_parts(array.as_ptr().cast(), len))
+    }
+}
+
+/// Build time
+pub const BUILD_TIME: &str = env!("ESP_BOOTLOADER_SUPPORT_BUILD_TIME");
+
+/// Build date
+pub const BUILD_DATE: &str = env!("ESP_BOOTLOADER_SUPPORT_BUILD_DATE");
+
+/// MMU page size in bytes
+pub const MMU_PAGE_SIZE: u32 = {
+    let mmu_page_size =
+        esp_config::esp_config_str!("ESP_BOOTLOADER_SUPPORT_ESP_IDF_CONFIG_MMU_PAGE_SIZE")
+            .as_bytes();
+    match mmu_page_size {
+        b"8k" => 8 * 1024,
+        b"16k" => 16 * 1024,
+        b"32k" => 32 * 1024,
+        b"64k" => 64 * 1024,
+        _ => 64 * 1024,
+    }
+};
+
+/// This macro populates the application descriptor (see [EspAppDesc]) which is
+/// available as a static named `ESP_APP_DESC`
+///
+/// In most cases you can just use the no-arguments version of this macro.
+#[macro_export]
+macro_rules! esp_app_desc {
+    () => {
+        $crate::esp_app_desc!(
+            env!("CARGO_PKG_VERSION"),
+            env!("CARGO_PKG_NAME"),
+            $crate::BUILD_TIME,
+            $crate::BUILD_DATE,
+            "5.3.1",
+            (31 - u32::leading_zeros($crate::MMU_PAGE_SIZE)) as u8,
+            0,
+            u16::MAX
+        );
+    };
+
+    (
+     $version: expr,
+     $project_name: expr,
+     $build_time: expr,
+     $build_date: expr,
+     $idf_ver: expr,
+     $mmu_page_size: expr,
+     $min_efuse_blk_rev_full: expr,
+     $max_efuse_blk_rev_full: expr
+    ) => {
+        const ESP_APP_DESC_MAGIC_WORD: u32 = 0xABCD5432;
+
+        const fn str_to_cstr_array<const C: usize>(s: &str) -> [::core::ffi::c_char; C] {
+            let bytes = s.as_bytes();
+            if bytes.len() >= C {
+                assert!(true, "String is too long for the C-string field");
```

Comment:
Should this be `false`?

I suppose you could just move the `bytes.len() < C` into the assert. 😄

Reply:
actually it's useless - the compiler will understand the overflow even before we can assert anything - so I removed this

(I admi, initially t I pretty much just copied this code from _somewhere_)